### ISSUE TITLE
std::variant in a std::map declaration makes IntelliSense crash

### ISSUE THREAD

**Issue by etremel** [microsoft/vscode-cpptools#5584]
Issue Type: <b>Bug</b>

Writing a brace-initializer list for a std::map whose values are std::variant types will cause IntelliSense to crash and cease providing suggestions or error squiggles for the entire file containing the initializer list. Even after deleting the offending declaration, IntelliSense will not work for the file that used to contain it until VS Code is restarted.

Steps to reproduce the problem:

1. In VSCode with the C/C++ extension enabled, create a new cpp source file and enter the following code:
```
#include <map>
#include <variant>

struct Foo {
    int a;
};
struct Bar {
    int a;
    bool b;
};

int main(int argc, char** argv) {
    std::map<int, std::variant<Foo, Bar>> test = {
        {0, Foo{1}},
        {1, Bar{2, true}},
    };
    
    return 0;
}
```
2. On the blank line before "return 0", type `test.`
    
     Expected result: An IntelliSense pop-up menu appears showing member functions of std::map as potential completions.
 
    Actual result: Nothing happens.

3. Press enter, leaving the incomplete line of code `test.`.

    Expected result: A red error squiggle underlines the invalid syntax.

    Actual result: Nothing happens.

Extension version: 0.28.1
VS Code version: Code 1.45.1 (5763d909d5f12fe19f215cbfdd29a91c0fa9208a, 2020-05-14T08:27:22.494Z)
OS version: Linux x64 4.15.0-100-generic snap

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-3770 CPU @ 3.40GHz (8 x 2397)|
|GPU Status|2d_canvas: unavailable_software<br>flash_3d: disabled_software<br>flash_stage3d: disabled_software<br>flash_stage3d_baseline: disabled_software<br>gpu_compositing: disabled_software<br>multiple_raster_threads: enabled_on<br>oop_rasterization: disabled_off<br>protected_video_decode: disabled_off<br>rasterization: disabled_software<br>skia_renderer: disabled_off_ok<br>video_decode: disabled_software<br>viz_display_compositor: enabled_on<br>viz_hit_test_surface_layer: disabled_off_ok<br>webgl: unavailable_software<br>webgl2: unavailable_software|
|Load (avg)|0, 0, 0|
|Memory (System)|15.62GB (1.97GB free)|
|Process Argv|--no-sandbox|
|Screen Reader|no|
|VM|0%|
|DESKTOP_SESSION|awesome-gnome|
|XDG_CURRENT_DESKTOP|GNOME|
|XDG_SESSION_DESKTOP|awesome-gnome|
|XDG_SESSION_TYPE|x11|
</details>

**Comment by sean-mcmanus** [comment]
I'm not reproing the issue. Can you run C/C++: Log Diagnostics and let us know what compiler and compiler version, cpp standard version, and IntelliSense mode is being used? When you use Go to Def on map/variant, do they go to /usr/include/c++/10 paths (or what version instead of 10 is being used)?

**Comment by etremel** [comment]
I just reproduced the issue again on my computer. Here's the relevant part of the C/C++: Log Diagnostics output (I cut out some "translation unit configurations" for other unrelated files in my project):

```
-------- Diagnostics - 6/2/2020, 9:38:10 AM
Version: 0.28.2
Current Configuration:
{
    "name": "Linux",
    "includePath": [
        "${workspaceFolder}/include",
        "${workspaceFolder}/**"
    ],
    "defines": [
        "USE_VERBS_API",
        "DERECHO_DEBUG"
    ],
    "compilerPath": "/usr/bin/g++",
    "cStandard": "c11",
    "cppStandard": "c++17",
    "intelliSenseMode": "gcc-x64",
    "compilerArgs": [
        "-fPIC"
    ],
    "browse": {
        "path": [
            "${workspaceFolder}/include",
            "${workspaceFolder}/**"
        ],
        "limitSymbolsToIncludedHeaders": true
    }
}

Translation Unit Configurations:
...
[ /home/edward/research/eclipse-cdt-workspace/derecho/src/applications/tests/unit_tests/vscode_parser_test.cpp ]: not ready
    Process ID: 20076
    Memory Usage: 0 MB
    Compiler Path: /usr/bin/g++
    Includes:
        /home/edward/research/eclipse-cdt-workspace/derecho/include
        /usr/include/c++/7
        /usr/include/x86_64-linux-gnu/c++/7
        /usr/include/c++/7/backward
        /usr/lib/gcc/x86_64-linux-gnu/7/include
        /usr/local/include
        /usr/lib/gcc/x86_64-linux-gnu/7/include-fixed
        /usr/include/x86_64-linux-gnu
        /usr/include
    Defines:
        USE_VERBS_API
        DERECHO_DEBUG
    Standard Version: c++17
    IntelliSense Mode: gcc-x64
    Other Flags:
        --g++
        --gnu_version=70500
Total Memory Usage: 396 MB
```

I can confirm that if I ctrl-click on std::map and std::variant, VS Code jumps to the correct header in /usr/include/c++/7/.

**Comment by sean-mcmanus** [comment]
Thanks for the info -- I got a repro (looks like some code in the gcc-7 and 8 headers is causing our parser to crash, but it doesn't repro with gcc 10).

**Comment by sean-mcmanus** [comment]
This appears to be the same crash as https://github.com/microsoft/vscode-cpptools/issues/5481 (from the call stack). The additional repro is potentially helpful though. A simple repro is:
```
#include <variant>
std::variant<double, int> aaa[] = { 1.0, 1 };
```

**Comment by sean-mcmanus** [comment]
This should be fixed in our next 0.29.0-insiders release (not 0.28.3).

**Comment by sean-mcmanus** [comment]
This crash should be fixed with https://github.com/microsoft/vscode-cpptools/releases/tag/0.29.0-insiders .
